File: frontend/src/api/history.ts
import { BaseURL } from "./base.ts";

export enum Applications {
    infuse = "Infuse (com.firecore.infuse)",
    kinopub = "MicroIPTV (octavian.microiptv)",
    netflix = "Netflix (com.netflix.Netflix)",
    podcasts = "Overcast (com.apple.TVAirPlay)",
    twitch = "VLC (org.videolan.vlc-ios)",
    viaplay = "Viaplay (se.harbourfront.viasatondemand)",
    youtube = "YouTube (com.google.ios.youtube)",
}

export type HistoryItem = {
    application: Applications;
    artist: string;
    id: string;
    publishedAt: string;
    thumbnail: string;
    title: string;
    url: string;
};

export type Playback = {
    contentId: string;
    finishTime: string;
    id: string;
    startTime: string;
};

export type FullHistory = {
    content: Map<string, HistoryItem>;
    playback: Playback[];
};

export const getFullHistory = async (): Promise<FullHistory> => {
    const resp = await fetch(`${BaseURL}/api/history`, {
        method: "GET",
    });

    if (!resp.ok) {
        throw new Error("Failed to fetch history");
    }

    const data = await resp.json();

    const content = data.items.reduce((acc: Map<string, HistoryItem>, item: HistoryItem) => {
        acc.set(item.id, item);
        return acc;
    }, new Map());

    return {
        content,
        playback: data.playback,
    };
};

File: app/user/history.go
package user

import (
	"content-oracle/app/providers"
	"fmt"
	"log"
	"sort"
	"time"
)

type History struct {
	zimaClient *providers.Zima
	baseURL    string
}

func NewHistory(zimaClient *providers.Zima, baseURL string) *History {
	return &History{
		zimaClient: zimaClient,
		baseURL:    baseURL,
	}
}

type Item struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Arist       string `json:"artist"`
	Thumbnail   string `json:"thumbnail"`
	Url         string `json:"url"`
	PublishedAt string `json:"publishedAt"`
	Application string `json:"application"`
}

type PlaybackItem struct {
	ContentID  string    `json:"contentId"`
	StartTime  time.Time `json:"startTime"`
	FinishTime time.Time `json:"finishTime"`
}

type Playback struct {
	ContentID  string    `json:"contentId"`
	StartTime  time.Time `json:"startTime"`
	FinishTime time.Time `json:"finishTime"`
}

type FullHistory struct {
	Items    []Item     `json:"items"`
	Playback []Playback `json:"playback"`
}

func (p *History) GetAll() (*FullHistory, error) {
	fullHistory, err := p.zimaClient.GetContent(true, "")
	if err != nil {
		log.Printf("[ERROR] failed to get youtube history: %s", err)
		return nil, err
	}

	allPlayback := make([]providers.ZimaPlayback, 0)

	for _, item := range fullHistory {
		for _, playback := range item.Playback {
			allPlayback = append(allPlayback, playback)
		}
	}

	sort.Slice(allPlayback, func(i, j int) bool {
		return allPlayback[i].UpdatedAt > allPlayback[j].UpdatedAt
	})

	playback := make([]Playback, 0)
	for index, item := range allPlayback {
		updatedAt, err := time.Parse(time.RFC3339, item.UpdatedAt)
		if err != nil {
			log.Printf("[ERROR] failed to parse updated at time: %s", err)
			continue
		}

		if index == 0 || playback[len(playback)-1].ContentID != item.ContentID {
			playback = append(playback, Playback{
				ContentID:  item.ContentID,
				StartTime:  updatedAt,
				FinishTime: updatedAt,
			})
			continue
		}

		playback[len(playback)-1].StartTime = updatedAt
	}

	var history []Item

	for _, item := range fullHistory {
		historyItem := Item{
			ID:          item.ID,
			Title:       item.Title,
			Arist:       item.Artist,
			Application: item.Application,
		}

		if item.Metadata != nil {
			historyItem.Thumbnail = fmt.Sprintf("%s/api/proxy?url=%s", p.baseURL, item.Metadata.PosterLink)
			historyItem.Url = item.Metadata.ContentUrl
		}

		history = append(history, historyItem)
	}

	return &FullHistory{
		Playback: playback,
		Items:    history,
	}, nil
}
